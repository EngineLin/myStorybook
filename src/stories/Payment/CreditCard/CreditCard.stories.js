import CreditCardComponent from './CreditCard'

export default {
    title: 'Payment/CreditCard',
    component: CreditCardComponent,
    argTypes: {
        bankName: {
            control: {
                type: 'text',
            },
        },
        cardNumber: {
            control: {
                type: 'text',
            },
        },
        userName: {
            control: {
                type: 'text',
            },
        },
        deadline: {
            control: {
                type: 'text',
            },
        },
    }
}

export const Default = (args = {}) => ({
    components: {CreditCardComponent},
    props: Object.keys(args),
    template: `
      <CreditCardComponent 
          :bankName="bankName" 
          :cardNumber="cardNumber" 
          :userName="userName" 
          :deadline="deadline"
      />
    `,
})
Default.args = {
    bankName: 'VISA',
    cardNumber: '1233232194890398',
    userName: 'Engine Lin',
    deadline: '01/01'
}
