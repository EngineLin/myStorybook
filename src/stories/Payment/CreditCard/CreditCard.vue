<template>
  <div class="card
              w-auto max-w-sm
              flex flex-col justify-between
              px-8 py-4 border border-gray-500 rounded-lg"
  >
    <div class="header">
      <p class="header-text
                pb-3 pr-2
                font-bold tracking-wider text-right text-4xl text-blue-200">
        {{bankName}}</p>
    </div>

    <div class="chip-image-container
                relative h-12">
      <img src="../../../assets/images/icons/credit-card-chip.icon.png"
           class="chip-image
                  h-20 w-20"
           alt="chip">
    </div>

    <div class="card-number-container
                  flex justify-between align-center px-2 pt-5 pb-8">

      <!--v-for for 4 group of card number-->
      <p v-for="number in cardNumberArray"
         class="card-number text-outside-shadow text-2xl tracking-widest">
        {{ number }}</p>
    </div>

    <div class="footer
                flex justify-between align-center">
      <div class="user-name">
        <p class="text-xl text-white text-outside-shadow">
          {{userName}}</p>
      </div>

      <div class="deadline">
        <p class="text-lg text-white text-outside-shadow">
          {{deadline}}</p>
      </div>
    </div>
  </div>
</template>

<script>
  import {Component, Vue, Prop} from 'vue-property-decorator'

  @Component({props: ['bankName', 'cardNumber', 'userName', 'deadline']})
  export default class BasicCreditCard extends Vue {
    @Prop()
    bankName

    @Prop()
    cardNumber

    @Prop()
    userName

    @Prop()
    deadline

    get cardNumberArray() {
      const array = this.cardNumber.split('')

      return array
        .reduce((prov, curr, index) => {
          const order = Math.floor(index / 4)
          if (prov[order]) {
            prov[order].push(curr)
          } else {
            prov[order] = [curr]
          }
          return prov
        }, [])
        .map(array => array.join(''))
    }
  }
</script>

<style lang="scss" scoped>
  .card {
    background: linear-gradient(to right, #68a8ba, #347faa);
  }

  .chip-image-container {
    .chip-image {
      position: absolute;
      top: 50%;
      left: 10%;
      transform: translate(-50%, -50%);
    }
  }

  .text-outside-shadow {
    text-shadow: 0 -1px 2px #000000;
    color: #ffffff;
  }
</style>
