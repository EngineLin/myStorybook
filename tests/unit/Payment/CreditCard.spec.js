import { shallowMount } from '@vue/test-utils'

import CreditCard from '@/stories/Payment/CreditCard/CreditCard';

const initData = () => ({
    bankName: 'VISA',
    cardNumber: '1233232194890398',
    userName: 'Engine Lin',
    deadline: '01/01'
})

it('Match snapshot', () => {
    const wrapper = shallowMount(CreditCard, {
        propsData: initData()
    })
    expect(wrapper.html()).toMatchSnapshot()
})

it('Renders props.bankName with passed', () => {
    const data = initData()
    const wrapper = shallowMount(CreditCard, {
        propsData: {...data}
    })
    expect(wrapper.find('.header-text').text()).toEqual(data.bankName)

    const cardNumberWrapper = wrapper.findAll('.card-number')
    const cardNumbers = ['1233', '2321', '9489', '0398']

    expect(cardNumberWrapper.at(0).text()).toEqual(cardNumbers[0])
    expect(cardNumberWrapper.at(1).text()).toEqual(cardNumbers[1])
    expect(cardNumberWrapper.at(2).text()).toEqual(cardNumbers[2])
    expect(cardNumberWrapper.at(3).text()).toEqual(cardNumbers[3])

    expect(wrapper.find('.user-name p').text()).toEqual(data.userName)
    expect(wrapper.find('.deadline p').text()).toEqual(data.deadline)
})

it('Calculate computed cardNumberArray correctly.', async () => {
    const data = initData()
    const wrapper = shallowMount(CreditCard, {
        propsData: {...data}
    })
    await wrapper.setProps({ cardNumber: '1111222233334444' })
    expect(wrapper.vm.cardNumberArray).toEqual(['1111', '2222', '3333', '4444'])
})


