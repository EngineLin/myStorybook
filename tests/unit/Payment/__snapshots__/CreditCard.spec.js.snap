// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Match snapshot 1`] = `
"<div class=\\"card
            w-auto max-w-sm
            flex flex-col justify-between
            px-8 py-4 border border-gray-500 rounded-lg\\">
  <div class=\\"header\\">
    <p class=\\"header-text
              pb-3 pr-2
              font-bold tracking-wider text-right text-4xl text-blue-200\\">
      VISA</p>
  </div>
  <div class=\\"chip-image-container
              relative h-12\\"><img src=\\"../../../assets/images/icons/credit-card-chip.icon.png\\" alt=\\"chip\\" class=\\"chip-image
                h-20 w-20\\"></div>
  <div class=\\"card-number-container
                flex justify-between align-center px-2 pt-5 pb-8\\">
    <p class=\\"card-number text-outside-shadow text-2xl tracking-widest\\">
      1233</p>
    <p class=\\"card-number text-outside-shadow text-2xl tracking-widest\\">
      2321</p>
    <p class=\\"card-number text-outside-shadow text-2xl tracking-widest\\">
      9489</p>
    <p class=\\"card-number text-outside-shadow text-2xl tracking-widest\\">
      0398</p>
  </div>
  <div class=\\"footer
              flex justify-between align-center\\">
    <div class=\\"user-name\\">
      <p class=\\"text-xl text-white text-outside-shadow\\">
        Engine Lin</p>
    </div>
    <div class=\\"deadline\\">
      <p class=\\"text-lg text-white text-outside-shadow\\">
        01/01</p>
    </div>
  </div>
</div>"
`;
